### DISCUSSION TITLE
Switching between visual and source editor for qmd files

### DISCUSSION THREAD

**Discussion by Andrzej-Andrzej** [quarto-dev/quarto-cli#6027] [Q&A]
### Description

Hi,
1. when I switch between visual and source editor in RStudio the title of my file is changing in the tab to red colour so I need to save it to preserve changes. I did not do anything apart from switching between editors. Why is that ? Is it normal behaviour ? 
2.  is there any auto formatting turned on for qmd files ? I mean when I write :::: (4) colons down in source editor, after saving a file, closing and opening it again I have got ::: (3) instead of :::: (4) colons. Why is that ? 

I am on windows 10, quarto version 1.3.361, R version 4.3.0 patched, Rstudio 2023.08.0 Build 60.

**Comment by baptiste** [comment]
My understanding is that turning visual mode on [triggers a normalisation of the file's contents](https://github.com/quarto-dev/quarto-cli/discussions/3102), i.e. the markdown formatting is parsed and re-written in a standardised form. For instance definition lists written manually like this,

```
Item 1:
  ~ this is a line
  ~ this is another

Item 2:
  ~ this is a line
  ~ this is another
  
```

become, after toggling visual mode on and off:

```
Item 1:

:   this is a line

:   this is another

Item 2:

:   this is a line

:   this is another

```

**Reply by Andrzej-Andrzej** [reply]
Thank you baptiste, can this "formatting" be turned off somehow, but still I like to use visual editor in order to insert pictures, etc. ?

**Reply by mcanouil** [reply]
I don't believe you can, at least at this moment, use the visual editor and get non standardised output.

To note, this is an IDE feature that has basically nothing to do with Quarto.
You'll have better answers asking on the RStudio IDE repository: <https://github.com/rstudio/rstudio>.

**Reply by Andrzej-Andrzej** [reply]
I have tried it with VisualStudio Code and it's the same. VS Code has got visual editor for quarto as well. The effect is the same, syntax become corrected when going from visual to source editor and vice versa. So maybe this is not RStudio issue ?

**Reply by mcanouil** [reply]
Well, I know the visual editor is a common feature and that at some point the underlying code was supposed to be the same (_i.e._, "editor" part of https://github.com/quarto-dev/quarto).

I think @jjallaire is the best one to help identify where this report/feedback should go.

**Reply by jjallaire** [reply]
Yes, this is indeed more something that the Quarto team works on (me specifically) so its appropriate to follow up on visual editor issues here. The normalization isn't something that can be prevented (we used Pandoc to parse the markdown into the Pandoc AST, then we edit the AST, then back through Pandoc to markdown.

**Reply by moritzketzer** [reply]
I also have an issue with switching into visual editor and back.
When authoring markdown and using git to version control, it is helpful to do a linebreak after each sentence, in order to properly track changes.
The visual editor though, automatically formats sentences connected by linebreaks, i.e., a paragraph, into a single line.
The problem with this is, that whenever someone changes just a single word the entire paragraph gets highlighted and it is very difficult to actually track the changes that were made.
It would be nice to have the possibility of keeping the manual formatting with linebreaks between sentences or to have a setting that forces the visual editor to automatically format a paragraph as such, that each sentence is it's own line, and that paragraphs are split by using an empty line.

This would much improve a git version controlled collaborative writing workflow in quarto.

**Reply by cderv** [reply]
@moritzketzer there is a configuration that could help here: 
https://quarto.org/docs/visual-editor/vscode/#line-wrapping

**Reply by moritzketzer** [reply]
I did not see that, thank you for pointing this out :)

**Reply by monirtiktok938-a11y** [reply]
Item 1:

:   this is a line

:   this is another

Item 2:

:   this is a line

:   this is another

**Comment by n8layman** [comment]
I also have an issue. Whenever I add a list in the code editor and then switch to visual editor it swaps out the single space between the dash and list item from one to three spaces. So `- [ ] item 1` becomes `-   [ ] item 1` after using the visual editor.

**Reply by mcanouil** [reply]
The [Visual Editor]() uses Pandoc AST to normalise Markdown where several different syntax can mean the same things.
This is Pandoc doing the addition of the extra spaces for some reason:

```sh
echo "- [ ] Whatever" | quarto pandoc --from markdown --to markdown -
#> -   [ ] Whatever
```

- https://github.com/jgm/pandoc/issues/7172

```sh
echo "- [ ] Whatever" | quarto pandoc --from markdown --to commonmark_x -       
#> - [ ] Whatever
```

Note that the possible changes of formatting is stated in a modal dialogue that opens when you first use the Visual Editor:

> Markdown code will be re-formatted using the Pandoc markdown writer.

**Reply by n8layman** [reply]
I see that makes sense. Thanks!

**Reply by n8layman** [reply]
Is there a way to force the visual editor to use commonmark_x? Setting it in the document yaml doesn't seem to change the behavior.

**Reply by mcanouil** [reply]
No, there is not currently and I doubt there will be in a near future because this will lead to other issues.

FYI, the repository with the Visual Editor is https://github.com/quarto-dev/quarto

**Reply by mcanouil** [reply]
@n8layman Regarding the spaces in list, this is going to change in the future (few months I would say) as this was just fixed in Pandoc.

- https://github.com/jgm/pandoc/issues/7172

**Reply by n8layman** [reply]
Oh great!
